### PR TITLE
Improved HttpContext pooling

### LINKED ISSUE #11069: [Discussion] NullReferenceException at Microsoft.AspNetCore.Http.Internal.DefaultHttpRequest.get_Query()
This topic came from several other:
#2806
#3219
#3219

They all a bit different, but the reason (I think) the same: concurent access to HttpContext from different threads.

Our story is pretty similar: we have project on asp.net 1.1, and we widely using HttpContext and HttpContextContextAccessor for different things (mainly for custom authorization and access to Session info), and I'm not sure if all what we're doing is really safe.

My first question is: there was a several mentions in issues above that middleware is kind of safe place, but what about ActionFilters? 
We have a pretty easy one like this:
```
public class PartialViewFilter : IActionFilter
    {
        public void OnActionExecuted(ActionExecutedContext context)
        {
            
        }

        public void OnActionExecuting(ActionExecutingContext context)
        {
            context.HttpContext.Items["_partial"] = context.HttpContext.Request.Query.Keys.Contains("partial");
        }
    }
```
and catches NullReferenceException sometimes (not very often, less than 5% requests aprox).
Exception:

> System.NullReferenceException: Object reference not set to an instance of an object.
> at Microsoft.AspNetCore.Http.Internal.DefaultHttpRequest.get_Query()
> at ****.PartialViewFilter.OnActionExecuting
> …

The most interesting part is that problem start to be occurs only when we moved from 1.0 to 1.1 (no other changes at all). 

And now my second question: is any Best Practices how to avoid this? Or at least how to detect such things?

And the last one; what about to generate more accurate error? NullReferenceException is not really nice.

### PR DESCRIPTION
**Pull request by benaadams**
Improved HttpContext pooling from https://github.com/aspnet/AspNetCore/pull/10319

Moves pooling from Kestrel to HttpContextFactory

Resolves #11069

/cc @davidfowl

### CODE REVIEW

**Comment by benaadams** [comment]
Rebased

**Comment by benaadams** [comment]
Rebased

**Comment by benaadams** [comment]
egh; something weird happened with the rebase

**Comment by benaadams** [comment]
Fixed rebase

File: src/Http/Http.Features/src/FeatureReferences.cs [hunk 1]
```diff
@@ -108,5 +108,17 @@ public void Initalize(IFeatureCollection collection, int revision)
 
         public TFeature Fetch<TFeature>(ref TFeature cached, Func<IFeatureCollection, TFeature> factory)
             where TFeature : class => Fetch(ref cached, Collection, factory);
+
+        private static int ContextDisposed()
```
Comment (dougbu) [inline src/Http/Http.Features/src/FeatureReferences.cs]: Curious: Why add these extra methods instead of changing their use above to
``` c#
if (Collection?.Revision == null)
{
    throw new ObjectDisposedException("HttpContext", "Request has finished and HttpContext disposed.");
}
```

File: src/Http/Http/src/DefaultHttpContext.cs [hunk 2]
```diff
@@ -90,7 +90,7 @@ public void Uninitialize()
         private IHttpRequestIdentifierFeature RequestIdentifierFeature =>
             _features.Fetch(ref _features.Cache.RequestIdentifier, _newHttpRequestIdentifierFeature);
 
-        public override IFeatureCollection Features => _features.Collection;
+        public override IFeatureCollection Features => _features.Collection ?? ContextDisposed();
```
Comment (dougbu) [inline src/Http/Http/src/DefaultHttpContext.cs]: Similarly, why not use a single body here instead of two methods?
```suggestion
        public override IFeatureCollection Features
        {
            get
            {
                if (_features.Collection == null)
                {
                    throw new ObjectDisposedException(nameof(HttpContext), $"Request has finished and {nameof(HttpContext)} disposed.");
                }

                return _features.Collection;
            }
        }
```

**Review by dougbu — COMMENTED** [review]
@benaadams realize my review request was due to a bad rebase. But, a couple of curiosity / style questions&hellip;

File: src/Http/Http.Features/src/FeatureReferences.cs (hunk 1 above)
Comment (benaadams) [reply src/Http/Http.Features/src/FeatureReferences.cs]: Inlining. 

`Load string, Load string, New Exception, Throw` is quite a chunky bit of code to inline to every caller

File: src/Http/Http/src/DefaultHttpContext.cs (hunk 2 above)
Comment (benaadams) [reply src/Http/Http/src/DefaultHttpContext.cs]: The Jit has special optimizations for a void method that only throws.

1. It recognises it as cold code (shouldn't be throwing exceptions in common paths), so generates the branches for it to be unpredicted (also does this with regular exception throws)
2. As it recognised it as cold code it won't inline it; whereas regular throws are not split out (Jit doesn't split methods); so the instruction cache is hotter (exception throws are quite big).
3. It doesn't generate the asm for it until the exception is thrown (faster Jitting).
4. As it recognises it as non-returning, it doesn't bother trying to sort out the stack after the call (popping regsiters back), meaning the code is hotter.
5. If the method is generic, moving the throws into a non-returning method in a non-generic class drastically reduces the asm size generated (doesn't do it for every type).
6. If the method inlines, moving the the throws into a non-returning method reduces the asm size generated as it doesn't regenerate it for every caller.

tl;dr makes the CPU instruction cache better; generics smaller; inlines smaller; faster Jitting

**Comment by benaadams** [comment]
Superseded by https://github.com/aspnet/AspNetCore/pull/12385

**Comment by benaadams** [comment]
Superseded by https://github.com/aspnet/AspNetCore/pull/12385
